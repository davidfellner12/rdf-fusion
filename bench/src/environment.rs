use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the current directory. This might be different from the `data_dir` if a
    /// benchmark is running.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    pub fn push_dir(&self, dir: &str) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(dir);
        results_dir.push(dir);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(&benchmark_name.dir_name())?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
