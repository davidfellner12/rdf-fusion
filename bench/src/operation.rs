use crate::runs::BenchmarkRun;
use futures::StreamExt;
use rdf_fusion::store::Store;
use rdf_fusion::{Query, QueryExplanation, QueryOptions, QueryResults};

#[derive(Clone)]
pub enum SparqlRawOperation<QueryName> {
    Query(QueryName, String),
}

impl<QueryName: Clone> SparqlRawOperation<QueryName> {
    pub fn parse(&self) -> anyhow::Result<SparqlOperation<QueryName>> {
        match self {
            SparqlRawOperation::Query(query_name, query) => {
                let query = query.parse()?;
                Ok(SparqlOperation::Query(query_name.clone(), query))
            }
        }
    }
}

#[derive(Clone)]
pub enum SparqlOperation<QueryName> {
    Query(QueryName, Query),
}

impl<QueryName> SparqlOperation<QueryName> {
    pub fn query(&self) -> &Query {
        match self {
            SparqlOperation::Query(_, query) => query,
        }
    }

    pub async fn run(
        &self,
        store: &Store,
    ) -> anyhow::Result<(BenchmarkRun, QueryExplanation)> {
        let start = datafusion::common::instant::Instant::now();

        let options = QueryOptions::default();
        let explanation = match &self {
            SparqlOperation::Query(_, q) => {
                let (result, explanation) =
                    store.explain_query_opt(q.clone(), options.clone()).await?;
                match result {
                    QueryResults::Boolean(_) => (),
                    QueryResults::Solutions(s) => {
                        let mut stream = s.into_record_batch_stream()?;
                        while let Some(s) = stream.next().await {
                            s?;
                        }
                    }
                    QueryResults::Graph(mut g) => {
                        while let Some(t) = g.next().await {
                            t?;
                        }
                    }
                }
                explanation
            }
        };

        let duration = start.elapsed();
        Ok((BenchmarkRun { duration }, explanation))
    }
}

impl<QueryName: Clone> SparqlOperation<QueryName> {
    pub fn query_name(&self) -> QueryName {
        match self {
            SparqlOperation::Query(name, _) => name.clone(),
        }
    }
}
