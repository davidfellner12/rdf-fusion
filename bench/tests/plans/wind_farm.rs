use crate::plans::canonicalize_uuids;
use anyhow::Context;
use datafusion::physical_plan::displayable;
use datafusion::prelude::SessionConfig;
use insta::assert_snapshot;
use rdf_fusion::store::Store;
use rdf_fusion::{QueryExplanation, QueryOptions};
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::benchmarks::windfarm::{
    NumTurbines, WindFarmBenchmark, WindFarmQueryName, get_wind_farm_raw_sparql_operation,
};
use rdf_fusion_bench::environment::{BenchmarkContext, RdfFusionBenchContext};
use rdf_fusion_bench::operation::SparqlRawOperation;
use std::path::PathBuf;

#[tokio::test]
pub async fn optimized_logical_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        assert_snapshot!(
            format!("{name} (Optimized)"),
            canonicalize_uuids(&explanation.optimized_logical_plan.to_string())
        )
    })
    .await;
}

#[tokio::test]
pub async fn execution_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        let string = displayable(explanation.execution_plan.as_ref())
            .indent(false)
            .to_string();
        assert_snapshot!(
            format!("{name} (Execution Plan)"),
            canonicalize_uuids(&string)
        )
    })
    .await;
}

async fn for_all_explanations(assertion: impl Fn(String, QueryExplanation) -> ()) {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"));

    // Load the benchmark data and set max query count to one.
    let benchmark = WindFarmBenchmark::new(NumTurbines::N4);
    let benchmark_name = benchmark.name();
    let benchmark_context = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store =
        Store::new_with_datafusion_config(SessionConfig::new().with_target_partitions(1));
    for query_name in WindFarmQueryName::list_queries() {
        let benchmark_name = format!("Wind Farm - {query_name}");
        let query = get_query_to_execute(&benchmark_context, query_name);

        let (_, explanation) = store
            .explain_query_opt(query.text(), QueryOptions::default())
            .await
            .unwrap();

        assertion(benchmark_name, explanation);
    }
}

fn get_query_to_execute(
    benchmark_context: &BenchmarkContext,
    query_name: WindFarmQueryName,
) -> SparqlRawOperation<WindFarmQueryName> {
    get_wind_farm_raw_sparql_operation(benchmark_context, query_name)
        .context("Could not list raw operations for Wind Farm benchmark. Have you prepared a wind-farm-4 dataset?")
        .unwrap()
}
