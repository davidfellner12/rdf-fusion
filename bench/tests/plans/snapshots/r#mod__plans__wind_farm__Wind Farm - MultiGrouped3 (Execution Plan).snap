---
source: bench/tests/plans/wind_farm.rs
expression: "&string"
---
ProjectionExec: expr=[ENC_PT(site_label@0) as site_label, ENC_PT(wtur_label@1) as wtur_label, year@2 as year, month@3 as month, day@4 as day, hour@5 as hour, minute_10@6 as minute_10, ENC_PT(<uuid>@7) as val_prod_avg, ENC_PT(<uuid>@8) as val_dir_avg, ENC_PT(<uuid>@9) as val_speed_avg]
  AggregateExec: mode=Single, gby=[site_label@1 as site_label, wtur_label@3 as wtur_label, ENC_PT(year@25) as year, ENC_PT(month@24) as month, ENC_PT(day@23) as day, ENC_PT(hour@22) as hour, ENC_PT(minute_10@21) as minute_10], aggr=[<uuid>, <uuid>, <uuid>]
    CoalesceBatchesExec: target_batch_size=8192
      FilterExec: EBV(BOOLEAN_AS_TERM(EBV(BOOLEAN_AS_TERM(EBV(EQ(ENC_TV(site_label@1), ENC_TV({term_type:2,value:Wind Mountain,data_type:http://www.w3.org/2001/XMLSchema#string,language_tag:}))) AND EBV(GEQ(ENC_TV(t@14), ENC_TV({term_type:2,value:2022-08-30T08:46:53,data_type:http://www.w3.org/2001/XMLSchema#dateTime,language_tag:}))))) AND EBV(LEQ(ENC_TV(t@14), ENC_TV({term_type:2,value:2022-08-30T21:46:53,data_type:http://www.w3.org/2001/XMLSchema#dateTime,language_tag:})))))
        ProjectionExec: expr=[site@0 as site, site_label@1 as site_label, wtur_asp@2 as wtur_asp, wtur_label@3 as wtur_label, wtur@4 as wtur, gensys_asp@5 as gensys_asp, gensys@6 as gensys, generator_asp@7 as generator_asp, generator@8 as generator, weather_asp@9 as weather_asp, weather@10 as weather, ts_speed@11 as ts_speed, dp_speed@12 as dp_speed, val_speed@13 as val_speed, t@14 as t, ts_dir@15 as ts_dir, dp_dir@16 as dp_dir, val_dir@17 as val_dir, ts_prod@18 as ts_prod, dp_prod@19 as dp_prod, val_prod@20 as val_prod, MUL(ENC_TV({term_type:2,value:10,data_type:http://www.w3.org/2001/XMLSchema#integer,language_tag:}), FLOOR(DIV(MINUTES(ENC_TV(t@14)), ENC_TV({term_type:2,value:10.0,data_type:http://www.w3.org/2001/XMLSchema#decimal,language_tag:})))) as minute_10, HOURS(ENC_TV(t@14)) as hour, DAY(ENC_TV(t@14)) as day, MONTH(ENC_TV(t@14)) as month, YEAR(ENC_TV(t@14)) as year]
          CoalesceBatchesExec: target_batch_size=8192
            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp_prod@19, dp_prod@0), (t@14, t@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@17, ts_prod@18, dp_prod@19, val_prod@20]
              CoalesceBatchesExec: target_batch_size=8192
                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp_prod@19, dp_prod@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@17, ts_prod@18, dp_prod@19, val_prod@21]
                  CoalesceBatchesExec: target_batch_size=8192
                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts_prod@18, ts_prod@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@17, ts_prod@18, dp_prod@20]
                      CoalesceBatchesExec: target_batch_size=8192
                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts_prod@18, ts_prod@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@17, ts_prod@18]
                          CoalesceBatchesExec: target_batch_size=8192
                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(generator@8, generator@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@17, ts_prod@19]
                              CoalesceBatchesExec: target_batch_size=8192
                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts_dir@15, ts_dir@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@17]
                                  CoalesceBatchesExec: target_batch_size=8192
                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp_dir@16, dp_dir@0), (t@14, t@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@17]
                                      CoalesceBatchesExec: target_batch_size=8192
                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp_dir@16, dp_dir@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@16, val_dir@18]
                                          CoalesceBatchesExec: target_batch_size=8192
                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts_dir@15, ts_dir@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@15, dp_dir@17]
                                              CoalesceBatchesExec: target_batch_size=8192
                                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(weather@10, weather@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14, ts_dir@16]
                                                  CoalesceBatchesExec: target_batch_size=8192
                                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts_speed@11, ts_speed@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@14]
                                                      CoalesceBatchesExec: target_batch_size=8192
                                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp_speed@12, dp_speed@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@13, t@15]
                                                          CoalesceBatchesExec: target_batch_size=8192
                                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(dp_speed@12, dp_speed@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@12, val_speed@14]
                                                              CoalesceBatchesExec: target_batch_size=8192
                                                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(ts_speed@11, ts_speed@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@11, dp_speed@13]
                                                                  CoalesceBatchesExec: target_batch_size=8192
                                                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(weather@10, weather@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10, ts_speed@12]
                                                                      CoalesceBatchesExec: target_batch_size=8192
                                                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(weather@10, weather@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10]
                                                                          CoalesceBatchesExec: target_batch_size=8192
                                                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(weather_asp@9, weather_asp@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@9, weather@10]
                                                                              CoalesceBatchesExec: target_batch_size=8192
                                                                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur@4, wtur@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8, weather_asp@10]
                                                                                  CoalesceBatchesExec: target_batch_size=8192
                                                                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(generator@8, generator@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8]
                                                                                      CoalesceBatchesExec: target_batch_size=8192
                                                                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(generator_asp@7, generator_asp@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@7, generator@8]
                                                                                          CoalesceBatchesExec: target_batch_size=8192
                                                                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(gensys@6, gensys@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6, generator_asp@8]
                                                                                              CoalesceBatchesExec: target_batch_size=8192
                                                                                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(gensys@6, gensys@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6]
                                                                                                  CoalesceBatchesExec: target_batch_size=8192
                                                                                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(gensys_asp@5, gensys_asp@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@5, gensys@6]
                                                                                                      CoalesceBatchesExec: target_batch_size=8192
                                                                                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur@4, wtur@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4, gensys_asp@6]
                                                                                                          CoalesceBatchesExec: target_batch_size=8192
                                                                                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur@4, wtur@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4]
                                                                                                              CoalesceBatchesExec: target_batch_size=8192
                                                                                                                HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur_asp@2, wtur_asp@1)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@3, wtur@4]
                                                                                                                  CoalesceBatchesExec: target_batch_size=8192
                                                                                                                    HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(wtur_asp@2, wtur_asp@0)], projection=[site@0, site_label@1, wtur_asp@2, wtur_label@4]
                                                                                                                      CoalesceBatchesExec: target_batch_size=8192
                                                                                                                        HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(site@0, site@0)], projection=[site@0, site_label@1, wtur_asp@3]
                                                                                                                          CoalesceBatchesExec: target_batch_size=8192
                                                                                                                            HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(site@0, site@0)], projection=[site@0, site_label@2]
                                                                                                                              DataSourceExec: [GPOS] subject=?site, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#Site>
                                                                                                                              DataSourceExec: [GPOS] subject=?site, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object=?site_label, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                                                          DataSourceExec: [GPOS] subject=?site, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect>, object=?wtur_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                                                      DataSourceExec: [GPOS] subject=?wtur_asp, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object=?wtur_label, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                                                  DataSourceExec: [GPOS] subject=?wtur, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode>, object=?wtur_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                                              DataSourceExec: [GPOS] subject=?wtur, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#A>, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                                          DataSourceExec: [GPOS] subject=?wtur, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect>, object=?gensys_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                                      DataSourceExec: [GPOS] subject=?gensys, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode>, object=?gensys_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                                  DataSourceExec: [GPOS] subject=?gensys, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#RA>, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                              DataSourceExec: [GPOS] subject=?gensys, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect>, object=?generator_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                          DataSourceExec: [GPOS] subject=?generator, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode>, object=?generator_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                      DataSourceExec: [GPOS] subject=?generator, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#GAA>, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                                  DataSourceExec: [GPOS] subject=?wtur, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect>, object=?weather_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                              DataSourceExec: [GPOS] subject=?weather, predicate=<https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode>, object=?weather_asp, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                          DataSourceExec: [GPOS] subject=?weather, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<https://github.com/magbak/chrontext/rds_power#LE>, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                      DataSourceExec: [GPOS] subject=?weather, predicate=<https://github.com/magbak/chrontext#hasTimeseries>, object=?ts_speed, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                                  DataSourceExec: [GPOS] subject=?ts_speed, predicate=<https://github.com/magbak/chrontext#hasDataPoint>, object=?dp_speed, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                              DataSourceExec: [GPOS] subject=?dp_speed, predicate=<https://github.com/magbak/chrontext#hasValue>, object=?val_speed, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                          DataSourceExec: [GPOS] subject=?dp_speed, predicate=<https://github.com/magbak/chrontext#hasTimestamp>, object=?t, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                      DataSourceExec: [GPOS] subject=?ts_speed, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object="Windspeed", additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                                  DataSourceExec: [GPOS] subject=?weather, predicate=<https://github.com/magbak/chrontext#hasTimeseries>, object=?ts_dir, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                              DataSourceExec: [GPOS] subject=?ts_dir, predicate=<https://github.com/magbak/chrontext#hasDataPoint>, object=?dp_dir, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                          DataSourceExec: [GPOS] subject=?dp_dir, predicate=<https://github.com/magbak/chrontext#hasValue>, object=?val_dir, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                      DataSourceExec: [GPOS] subject=?dp_dir, predicate=<https://github.com/magbak/chrontext#hasTimestamp>, object=?t, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                                  DataSourceExec: [GPOS] subject=?ts_dir, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object="WindDirection", additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                              DataSourceExec: [GPOS] subject=?generator, predicate=<https://github.com/magbak/chrontext#hasTimeseries>, object=?ts_prod, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                          DataSourceExec: [GPOS] subject=?ts_prod, predicate=<http://www.w3.org/2000/01/rdf-schema#label>, object="Production", additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                      DataSourceExec: [GPOS] subject=?ts_prod, predicate=<https://github.com/magbak/chrontext#hasDataPoint>, object=?dp_prod, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
                  DataSourceExec: [GPOS] subject=?dp_prod, predicate=<https://github.com/magbak/chrontext#hasValue>, object=?val_prod, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
              DataSourceExec: [GPOS] subject=?dp_prod, predicate=<https://github.com/magbak/chrontext#hasTimestamp>, object=?t, additional_filters=[DynamicFilterPhysicalExpr [ true ]]
